themes/src/main/resources/theme/base/login/resources/js/passwordEntropy.js

```diff
@@ -0,0 +1,42 @@
+const clamp = (number, min, max) => Math.max(min, Math.min(number, max));
+
+const calculateEntropy = (password) => {
+    let charSetSize = 0;
+    if (/[a-z]/.test(password)) charSetSize += 26;
+    if (/[A-Z]/.test(password)) charSetSize += 26;
+    if (/[0-9]/.test(password)) charSetSize += 10;
+    if (/\W|_/.test(password)) charSetSize += 32; // 149_878 available unicode chars
+    return Math.log2(charSetSize ** password.length);
+}
+
+const checkEntropy = (entropyElement, password) =>  {
+    if (!password.length) {
+        entropyElement.style.display = "none";
+        return;
+    }
+
+    const entropyStatus = entropyElement.children.item(0).children.item(0);
+
+    // entropy source: https://www.baeldung.com/cs/password-entropy
+    const entropy = calculateEntropy(password);
+
+    // 00-24    poor password
+    // 25-49    weak password
+    // 50-74    good password
+    // 75-100   strong password
+    const rating = [entropyStatus.dataset.passwordPoor, entropyStatus.dataset.passwordWeak, entropyStatus.dataset.passwordGood, entropyStatus.dataset.passwordStrong];
+    const ratingRange = 25;
+    const ratingIndex = Math.trunc(entropy / ratingRange);
+    const ratingClassNames = ["progress-bar-danger", "progress-bar-warning", "progress-bar-info", "progress-bar-success"];
+    const clamped = clamp(ratingIndex, 0, rating.length - 1);
+
+    ratingClassNames.forEach(name => entropyStatus.classList.remove(name));
+    entropyStatus.innerText = rating[clamped];
+    entropyStatus.classList.add(ratingClassNames[clamped]);
+    entropyStatus.style.width = `${(clamped+1) * ratingRange}%`;
+    entropyElement.style.display = "block";
+}
+
+const entropyElement = document.querySelector('#password-entropy');
```

Comment:
Rather than using the `id` of the element I think we should be [doing the same](https://github.com/keycloak/keycloak/blob/ca2c5f688f03c0674eafd3368b93ab0762ad9b86/themes/src/main/resources/theme/base/login/resources/js/passwordVisibility.js#L14-L15) as we do for password visibility. Which is to use the `data` attribute to find the element on the page, and then bootstrapping all of the logic on that element.

Alternatively, since we're already using modern JavaScript, we could go the extra mile and just splurge on making this a full-fledged [Web Component](https://developer.mozilla.org/en-US/docs/Web/API/Web_components).

Thinking about it a bit more, I think Web Components might be a lot better of a harness than the sprinkles of JavaScript we've been doing. Also from a maintainability perspective.

WDYT?